### ISSUE TITLE
V2.7.8 - Invalid Daily Limit Error

### ISSUE THREAD

**Issue by mrmascari** [phw198/OutlookGoogleCalendarSync#687]
**OGCS Version**: v2.7.8
**Installed or Portable**: Portable

**Problem Description**  
Consistently generates a Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup. [403] error.

Using my own credentials (because everyone should), validated that quota limits are set to the max.  Everything works fine with 2.7.7.14.

Tried creating a new Google Credential and updating client information to use it.
Tried creating a new Google Calendar with no prior data.

Portable zip file download, extracted to it's own new folder, set up from scratch with new credential and google calendar and it still generated the error.

Is there a new API that needs to be added to the Google Project for the credential maybe?

The login screen to give it access to my Google mail account correctly shows my personal google project name that matches the credential.  It retrieves the calendar list fine, and is able to compare entries.

**Steps to Reproduce the Issue**  
1. Fresh Install
2. Run Sync

**Any other information**  
[OGcalsync-2.7.8.EDIT.log](https://github.com/phw198/OutlookGoogleCalendarSync/files/2992910/OGcalsync-2.7.8.EDIT.log)

```
2019-03-21 11:46:54,492 FINE  15 OutlookGoogleCalendarSync.Sync.Engine [0] -  Processing 30/445
2019-03-21 11:46:54,504 FINE  15 OutlookGoogleCalendarSync.OutlookOgcs.Calendar [0] -  GetSummary - not all day event
2019-03-21 11:46:54,505 FINE  15 OutlookGoogleCalendarSync.Recurrence [0] -  Found a master Outlook recurring item outside sync date range: 1/17/2019 1:00 PM (R) => "Some Meeting Title"
2019-03-21 11:46:54,608 DEBUG  15 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [0] -  Retrieving all events from Google: 1/17/2019 -> 1/18/2019
2019-03-21 11:46:54,657 WARN  15 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [0] -  Google.Apis.Requests.RequestError
Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup. [403]
Errors [
	Message[Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup.] Location[ - ] Reason[dailyLimitExceededUnreg] Domain[usageLimits]
]
```

**Comment by mrmascari** [comment]
Is it possible to get some extra debugging logging added to track down this issue?

On 2.7.7.14, this seems to happen when I first start the application.  Disconnecting the account and reconnecting to Google fixes it.  As long as I leave it running then, it'll keep working for many syncs.

On 2.7.8, it fails every time.

Somewhere between OutlookCalendar.cs in GetEventSummary, Recurrence.cs in GetGoogleMasterEvent, and GoogleCalendar.cs GetCalendarEntriesInRecurrence.

It's consistent that it runs for a while, generates lots of debug logging that it's processing and accessing google just fine.  Then generates the debug message "Retrieving all events from Google" to find the original of a repeating that is outside the range.  Followed bye the exception.
```
OutlookGoogleCalendarSync.GoogleOgcs.Calendar [0] -  Google.Apis.Requests.RequestError
Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup. [403]
Errors [
	Message[Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup.] Location[ - ] Reason[dailyLimitExceededUnreg] Domain[usageLimits]
]
```
And then the message from GoogleCalendar.cs handleAPIlimits.

I'm using my own Client ID specifically not to add load to the shared one, and as the only user shouldn't be anywhere near actual capacity. (Plus, it's the right thing to do.)  It feels like the error is something else, something where it's not actually a capacity limit, but something else hiding in there.  Just need some additional information.

I don't have a Visual Studio, or I would try to look myself.  :(

Thanks.

**Comment by phw198** [comment]
Do you have the Google+ API still enabled in your project? If not, try turning it on...

**Comment by mrmascari** [comment]
Yes, I have these 4 included in the project.

Google Calendar API
CalDAV API
Cloud Identity-Aware Proxy API
Google+ API


Only the Google Calendar API shows requests in the Google console dashboard.  The others were added as guesses at something that might show some details.

On 2.7.7.14 it flakes out and fails now and then.  Normally, disconnecting and reconnecting solves this.  It almost always requires this when first starting the application.  So, I normally leave it running for weeks to avoid this.  Sometimes it needs a few disconnect/reconnect to fix it.

On 2.7.8, it always fails first time.  I tried creating a new credential (same project), new empty google calendar, and setting up 2.7.8 from scratch without the prior settings file.  Still happens.

For both cases, it's always after many many successful Google requests.  Which is what makes the error message so confusing.  The error says it's Unauthenticated, but clearly it's been making authenticated requests just fine for hundreds of other calendar entries.

That's why I tried to track down where the error was coming from, but without an actual debugger (my fault), I get lost backtracking through some instance values.  It's always after "Retrieving all events from Google" looking for an earlier source of a recurring outside of range.  If you can add some extra error messages or other information after that call fails the next time you're building a hotfix/alpha after 2.7.8, I'll try it again with both the current and fresh scenarios, see if I can get some better debugging information.  Right now, the error that it's Unauthenticated feels like a misdirection hiding some other root cause.

**Comment by phw198** [comment]
To be honest, I have no idea :slightly_frowning_face: 

The error is happening on [this line](https://github.com/phw198/OutlookGoogleCalendarSync/blob/master/src/OutlookGoogleCalendarSync/GoogleOgcs/GoogleCalendar.cs#L245), which has been called many times before in your syncs without issue as you say. And there's nowhere to put extra logging - that call is a Google API call which fails and [falls into the handleApiLimits() function](https://github.com/phw198/OutlookGoogleCalendarSync/blob/master/src/OutlookGoogleCalendarSync/GoogleOgcs/GoogleCalendar.cs#L1562). None of that code has been changed in a long time.

If you do use the default API keys, does the problem go away? I.e. can it definitely be narrowed down to your own personal project?

**Comment by mrmascari** [comment]
Tried a fresh download of 2.7.9 and I didn't use my project credentials.  I also set it to a fresh empty Google calendar.  This way it would be a fresh run, creating all entries from Outlook into Google.

It give these messages:
```
2019-05-14 12:58:54,155 INFO  7 OutlookGoogleCalendarSync.Console [0] -  447 Outlook calendar entries found.
2019-05-14 12:58:54,551 INFO  7 OutlookGoogleCalendarSync.Console [0] -  0 Google calendar entries found.
2019-05-14 12:58:54,891 FINE  7 OutlookGoogleCalendarSync.Sync.Engine [0] -  Processing 447/447
2019-05-14 12:58:54,998 FINE  7 OutlookGoogleCalendarSync.Sync.Engine [0] -  Processing 446/447
2019-05-14 12:58:55,032 FINE  7 OutlookGoogleCalendarSync.Sync.Engine [0] -  Processing 445/447
......
2019-05-14 12:59:02,656 FINE  7 OutlookGoogleCalendarSync.Sync.Engine [0] -  Processing 29/447
2019-05-14 12:59:02,674 FINE  7 OutlookGoogleCalendarSync.Sync.Engine [0] -  Processing 28/447
2019-05-14 12:59:02,692 FINE  7 OutlookGoogleCalendarSync.Sync.Engine [0] -  Processing 27/447
2019-05-14 12:59:02,707 FINE  7 OutlookGoogleCalendarSync.OutlookOgcs.Calendar [0] -  GetSummary - not all day event
2019-05-14 12:59:02,708 FINE  7 OutlookGoogleCalendarSync.Recurrence [0] -  Found a master Outlook recurring item outside sync date range: 3/5/2019 2:00 PM (R) => "Tiering Check-point meeting "
2019-05-14 12:59:02,803 DEBUG  7 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [0] -  Retrieving all events from Google: 3/5/2019 -> 3/6/2019
2019-05-14 12:59:02,846 WARN  7 OutlookGoogleCalendarSync.GoogleOgcs.Calendar [0] -  Google.Apis.Requests.RequestError
Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup. [403]
Errors [
	Message[Daily Limit for Unauthenticated Use Exceeded. Continued use requires signup.] Location[ - ] Reason[dailyLimitExceededUnreg] Domain[usageLimits]
]
```
Full log attached too.

This is consistent with 2.7.8 and the intermittent error in 2.7.7.
[OGcalsync-Edit.log](https://github.com/phw198/OutlookGoogleCalendarSync/files/3179087/OGcalsync-Edit.log)

It's always after the "GetSummary - not all day event" message every time.

Thanks for the help too.

**Comment by phw198** [comment]
That's pretty inexplicable :confused: 

So we know:
* it's not your project or API keys
* it's not the default OGCS keys because no-one else is reporting this

So that leaves something to do with your computer/way you are accessing. Are you using a corporate proxy? My gut feel is some kind of geo-load balancing that is making Google drop your authentication because it doesn't trust you're still the person that started the calls. Pretty sure I saw this a long time ago where Google basically seemed to have black listed the user's IP for "abuse".

Have you specifically down-graded OGCS to v2.7.7 and the problem goes away? If it does, that probably debunks my proxy theory... Would be good if you can somehow test v2.7.9 without a proxy though.

**Comment by phw198** [comment]
Granted [this thread](https://stackoverflow.com/questions/42003125/user-rate-limit-exceeded-returned-for-only-one-user-in-domain-while-others-work) is talking about a different Google API and is a different rate-limiting error message, but it goes to the point of Google blacklisting your account. 

Could you try syncing with OGCS from the same computer but to another Google account?

And another [strange one](https://desk.draw.io/support/solutions/articles/16000045298-error-loading-file-daily-limit-for-unauthenticated-use-exceeded-continued-use-requires-signup-) - try clearing your browser (IE?) cache?

**Comment by mrmascari** [comment]
Yes, if I change back to 2.7.7.14 it works.

I frequently get the error when first starting 2.7.7.14.  But, disconnecting and reconnecting solves it every time.  Occasionally, 2.7.7.14 will also generate the error, but again disconnecting and reconnecting solves it every time.

I tried a bunch of different test scenarios today, between creating new credentials, creating a new calendar on the same account, creating a new project with new credentials.

I cleared the IE cache, and tried to look at the proxy settings.  Our IT has that setting locked out in IE and Chrome, but there is definitely a proxy somewhere since we get blocked site warnings.  It's possible it's a transparent one on the network and not in IE directly though.  I updated OGCS to not use the IE proxy settings.  I also tried while not logged in to VPN.

At some point in switching items and between versions, I was able to get 2.7.9 to start creating entries on a new calendar.  But, stopping it and restarting, it failed then.  No amount of disconnect/reconnect ever got it working again then.  I was never able to get it to try again then.

Between all the version switching, sometimes I get an OGCS error about a key already being added, not sure what that's related to.

It still feels like there is something else going on that the error message is misleading.  Either from google directly, or there's more than one error occurring and the second one is hiding the first.  Like a null pointer that hides an error doing something else first.  That's probably the worst part, since it has us looking at the wrong thing, or just trying stuff.  :(

When I created the new Google Project, I did notice that they work a little differently now then they did when I created the original.

I'm back on 2.7.7.14 and it's working now.

If you think of something else to try, I can give that a shot too.  I may give it a few days and try some of these options again to see if I can get one to work after some time.  Or, maybe I'll wait for the next Alpha and try out different options then.
